engine/access/state_stream/backend/handler.go

```diff
@@ -381,7 +381,7 @@ func (h *Handler) handleEventsResponse(send sendSubscribeEventsResponseFunc, hea
 			return status.Errorf(codes.Internal, "could not convert events to entity: %v", err)
 		}
 
-		index := messageIndex.Increment()
+		index := messageIndex.Value()
```

Comment:
nit: keep the get and set together so the stream fails quickly if there's an issue
```suggestion
		index := messageIndex.Value()
		if ok := messageIndex.Set(index + 1); !ok {
			return status.Errorf(codes.Internal, "message index already incremented to %d", messageIndex.Value())
		}
```